src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,695 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Represents a data structure in which each element additionally has a priority
+    /// associated with it.
+    /// </summary>
+    /// <typeparam name="TElement">The type of the element.</typeparam>
+    /// <typeparam name="TPriority">The type of the priority.</typeparam>
+    public class PriorityQueue<TElement, TPriority>
+    {
+        /// <summary>
+        /// Represents an implicit heap-ordered complete d-ary tree, stored as an array.
+        /// </summary>
+        private (TElement element, TPriority priority)[] _nodes;
+
+        private Lazy<UnorderedItemsCollection> _unorderedItems;
+
+        /// <summary>
+        /// The number of nodes in the heap.
+        /// </summary>
+        private int _size;
+
+        /// <summary>
+        /// Used to keep enumerator in sync with the collection.
+        /// </summary>
+        private int _version;
+
+        private const int MinimumGrow = 4;
+        private const int GrowFactor = 200; // double each time
+
+        private const int RootIndex = 0;
+
+        /// <summary>
+        /// Specifies the arity of the d-ary heap, which here is quaternary.
+        /// </summary>
+        private const uint Arity = 4;
+
+        /// <summary>
+        /// Creates an empty priority queue.
+        /// </summary>
+        public PriorityQueue()
+            : this(initialCapacity: 0, comparer: null)
+        {
+        }
+
+        /// <summary>
+        /// Creates an empty priority queue with the specified initial capacity for its underlying array.
+        /// </summary>
+        public PriorityQueue(int initialCapacity)
+            : this(initialCapacity, comparer: null)
+        {
+        }
+
+        /// <summary>
+        /// Creates an empty priority queue with the specified priority comparer.
+        /// </summary>
+        public PriorityQueue(IComparer<TPriority>? comparer)
+            : this(initialCapacity: 0, comparer)
+        {
+        }
+
+        /// <summary>
+        /// Creates an empty priority queue with the specified priority comparer and
+        /// the specified initial capacity for its underlying array.
+        /// </summary>
+        public PriorityQueue(int initialCapacity, IComparer<TPriority>? comparer)
+        {
+            if (initialCapacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(initialCapacity), initialCapacity, SR.ArgumentOutOfRange_NeedNonNegNum);
+            }
+
+            _nodes = new (TElement, TPriority)[initialCapacity];
```

Comment:
In the common case (e.g. `new PriorityQueue<..>()`), this is going to allocate an empty array.

Reply:
This is also how it behaves in `Stack<T>` for example — what would you suggest instead?

Reply:
> This is also how it behaves in Stack<T> for example

No,  it's not.  `new Stack<T>()` uses Array.Empty. But you've written the default ctor here to delegate to the one that always allocates

Reply:
You're right, for some cases `Stack<T>` uses `Array.Empty`, for some other cases it allocates an empty array. OK, we'll be better and never allocate one :D

Reply:
It uses Array.Empty when no check is needed because no capacity is specified.  If a capacity is specified, it's rarely zero and so typically doesn't warrant paying for a check. The simple ctor shouldn't delegate to the other ctors.